cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java

```diff
@@ -120,4 +228,87 @@ private HTable createHTable(TableId tableId) throws IOException {
     consumerTable.setAutoFlush(false);
     return consumerTable;
   }
+
+  /**
+   * A {@link QueueConsumer} that delegates to a list of consumers sequentially. It also has logic to renew
+   * the consumers list when all existing consumers has consumed everything in the current queue barrier.
+   */
+  private final class SmartQueueConsumer extends ForwardingTransactionAware implements QueueConsumer {
+
+    private final QueueName queueName;
+    private final ConsumerConfig consumerConfig;
+    private final Callable<? extends Iterable<HBaseQueueConsumer>> consumerCreator;
+    private final Deque<HBaseQueueConsumer> consumers;
+
+    private SmartQueueConsumer(QueueName queueName, ConsumerConfig consumerConfig,
+                               Callable<? extends Iterable<HBaseQueueConsumer>> consumerCreator) throws Exception {
+      this.queueName = queueName;
+      this.consumerConfig = consumerConfig;
+      this.consumers = Lists.newLinkedList(consumerCreator.call());
+      this.consumerCreator = consumerCreator;
+    }
+
+    @Override
+    protected TransactionAware delegate() {
+      return consumers.peek();
+    }
+
+    @Override
+    public QueueName getQueueName() {
+      return queueName;
+    }
+
+    @Override
+    public ConsumerConfig getConfig() {
+      return consumerConfig;
+    }
+
+    @Override
+    public DequeueResult<byte[]> dequeue() throws IOException {
+      return dequeue(1);
+    }
+
+    @Override
+    public DequeueResult<byte[]> dequeue(int maxBatchSize) throws IOException {
+      return consumers.peek().dequeue(maxBatchSize);
+    }
+
+    @Override
+    public void close() throws IOException {
+      for (HBaseQueueConsumer consumer : consumers) {
+        Closeables.closeQuietly(consumer);
+      }
+    }
+
+    @Override
+    public void startTx(Transaction tx) {
+      // This is a fallback case if we failed to update the consumers in the postTxCommit call.
+      if (consumers.isEmpty()) {
+        updateConsumers();
+      }
+      super.startTx(tx);
+    }
+
+    @Override
+    public void postTxCommit() {
+      super.postTxCommit();
+      HBaseQueueConsumer consumer = consumers.poll();
+      if (!consumer.isClosed()) {
+        consumers.add(consumer);
```

Comment:
So, a single consumer does not have multiple threads if it is catching up for an instance id that was removed. Instead this SmartConsumer round-robins through multiple instance ids. I am sure this simplifies the code... in terms of performance, though, catch-up to the barrier will take longer, right?